### ISSUE TITLE
v1.63.0 broken on Mac (bash 3.2)

### ISSUE THREAD

**Issue by Chris-Softrams** [antonbabenko/pre-commit-terraform#337]
### Describe the bug
https://github.com/antonbabenko/pre-commit-terraform/blob/v1.63.0/hooks/_common.sh#L30 uses `declare -g` which is not an option in all bash versions.

```bash
.cache/pre-commit/repo574zjemm/hooks/_common.sh: line 30: declare: -g: invalid option
declare: usage: declare [-afFirtx] [-p] [name[=value] ...]
```

Tried with 2 different versions of bash.

```
/bin/bash --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin20)
Copyright (C) 2007 Free Software Foundation, Inc.
```
and
```
/usr/local/bin/bash --version
GNU bash, version 5.1.16(1)-release (x86_64-apple-darwin20.6.0)
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

Version v1.62.3 works fine.

**Comment by Chris-Softrams** [comment]
@MaxymVlasov I even tried with https://formulae.brew.sh/formula/bash and setting my SHELL to `/usr/local/bin/bash`

**Comment by yermulnik** [comment]
> So looks like we need to maintain compatibility with bash 3.2 :roll_eyes:

Or include `Bash >= 4` as requirement in README =)

**Comment by yermulnik** [comment]
> @MaxymVlasov I even tried with https://formulae.brew.sh/formula/bash and setting my SHELL to `/usr/local/bin/bash`

You'd need to update your `PATH` var so that `/usr/local/bin` precedes `/usr/bin` (or which ever bash3 resides in).

**Comment by robinbowes** [comment]
> https://github.com/antonbabenko/pre-commit-terraform/blob/v1.63.0/hooks/_common.sh#L30 uses declare -g which is not an option in all bash versions.

That's relatively easy to fix.

`declare -g` declares a variable to be global when used inside a function. To work on Bash 3.2, simply move the variable declaration outside the function - eg, move to the top of the file. Arguably, that's clearer than declaring a global var in a function anyway.

Move this to the top of the file:

```
  # common global arrays.
  # Populated via `common::parse_cmdline` and can be used inside hooks' functions
  declare -a ARGS=() HOOK_CONFIG=() FILES=()
```

**Comment by ocofaigh** [comment]
Recently ran into the same issue recently, and came across this:

```
# For bash 4.x, must not be in posix mode, may use temporary files
mapfile -t array < <(mycommand)

# For bash 3.x+, must not be in posix mode, may use temporary files
array=()
while IFS='' read -r line; do array+=("$line"); done < <(mycommand)

# For ksh, and bash 4.2+ with the lastpipe option enabled (may require disabling monitor mode)
array=()
mycommand | while IFS="" read -r line; do array+=("$line"); done
```

I had to change my scripts to use option 2 instead of option 1 in order to be compatible on mac with enforcing a bash upgrade. Suggest to do the same here.

**Comment by drewmullen** [comment]
> ```yaml
> 48bc03ca3f0f2f782d2f430069868019a6892062
> ```

This worked for me, @MaxymVlasov . I was having the same error as the issue outlines when upgrading to from v1.62 to v1.64

**Comment by yermulnik** [comment]
Just to compile main items that are required to keep compatibility with Bash v3:
- builtins/utilities missing from Bash v3 or from base OS:
  - `decalare -g`: code changes → drop `declare -g` and (possibly) use `export` to populate var(s) globally where applicable and if needed
  - `mapfile`: fallback to `read -a` (looking at the code, where `mapfile` is used, we should be able to drop use of array at all (we only need the last item from that array, which can be accomplished by using `tail -1`)
  - `realpath`:
    - not enabled in Bash by default (requires `enable -f realpath realpath` to enable)
    - not built and enabled in Bash less than v4 by default
    - code changes:
      - bash version is equal or greater than 4 → enable this builtin
      - bash version is less than 4 and `realpath` binary is missing from `PATH` → implement `realpath` emulation via function and populate all shell hooks with it via e.g. `_common.sh`
        - alternatively can be "enabled" by installing `coreutils` using package manager on Linux and/or Homebrew on MacOS and Linux
     - is already in the list of requirements for MacOS along with Homebrew ([requirements installation steps for MacOS](https://github.com/antonbabenko/pre-commit-terraform#1-install-dependencies) assume `brew` is already installed)
     - is used by us in conjunction with `dirname`, which is part of `coreutils` at least on Ubuntu (is this available from base on other Linux distros and MacOS?)
      - alternatively use quirks like `cd -P -- "${PATH_TO_DIR_WITH_HOOK}" && pwd -P`
- all contributors are aware of restriction to use modern bash features which are not implemented in Bash v3 (http://tiswww.case.edu/php/chet/bash/NEWS) and of restriction to rely on common core packages like `coreutils`
- all contributors are aware of possible behavioral differences of common builtins and utilities (http://tiswww.case.edu/php/chet/bash/NEWS)
- all contributors have an option to test their code using Bash v3 (e.g. with respective Docker container — available at https://hub.docker.com/_/bash?tab=tags)
- all PRs must be automatically verified to be compatible with Bash v3 with e.g. GitHub workflow
- PR reviewers should bear in mind the above
- what else?

My only concerns are the items apart from the very first one and the need to support quite ancient version of shell interpreter.

Alternatively we can enforce Bash v4/v5 requirement. I realize this is intrusive for MacOS folks who are sort of enforced to use `pre-commit` (e.g. where corporate policy puts a requirement to use `pre-commit-terraform` for repos with Terraform code in the Company), though yet we already [require them to install a bunch of tools](https://github.com/antonbabenko/pre-commit-terraform#1-install-dependencies) and having one another relatively simple step like the below should be not a big deal:
```
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)" && \
    brew install bash && \
    # as far as I remember Homebrew on MacOS links binaries into /usr/local/bin
    export PATH="/usr/local/bin:${PATH+:$PATH}" && \
    echo 'PATH="/usr/local/bin:${PATH+:$PATH}"' >> ~/.bashrc
```

➡️ IMHO

**Comment by artis3n** [comment]
rev `48bc03ca3f0f2f782d2f430069868019a6892062` resolved this error for me on OSX 12.2.1

```yml
- repo: https://github.com/antonbabenko/pre-commit-terraform
   rev: 48bc03ca3f0f2f782d2f430069868019a6892062
   hooks:
     - id: terraform_fmt
     - id: terraform_docs
       args:
         - '--args=--sort-by required'
```

**Comment by sblack4** [comment]
I was able to get around this with a `brew install bash`. It looks like bash 5 is in Homebrew:
```
➜ brew info bash                                                                    
bash: stable 5.1.16 (bottled), HEAD
Bourne-Again SHell, a UNIX command interpreter
https://www.gnu.org/software/bash/
/usr/local/Cellar/bash/5.1.16 (157 files, 10.9MB) *
  Poured from bottle on 2022-03-07 at 15:14:01
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/bash.rb
License: GPL-3.0-or-later
==> Options
--HEAD
        Install HEAD version
==> Analytics
install: 25,150 (30 days), 100,366 (90 days), 331,470 (365 days)
install-on-request: 20,579 (30 days), 82,799 (90 days), 276,640 (365 days)
build-error: 33 (30 days)
```

**Comment by sms-astanley** [comment]
> ```yaml
> 48bc03ca3f0f2f782d2f430069868019a6892062
> ```

I tried this but still hitting this error.  Was using rev: v1.64.0 on MacOS 12.3
```
/Users/andrstan/.cache/pre-commit/repo0nc0qtfr/hooks/_common.sh: line 30: declare: -g: invalid option
declare: usage: declare [-afFirtx] [-p] [name[=value] ...]
```

**Comment by antonbabenko** [comment]
This issue has been resolved in version 1.64.1 :tada:
